### PR TITLE
Latejoin Job Selection

### PR DESCRIPTION
**Pull request by ike709**
Allows players to select a job when they latejoin.

Demo: https://streamable.com/dzw795

It won't let you join as a job if it has no open positions, but this is only conveyed to the player via a console message at the moment because I can't wrap my head around netcode. Ideally the buttons would be disabled, but like I said I don't know how to get the available positions from the server to the client.

### CODE REVIEW

**Comment by Zumorica** [comment]
Fix the conflicts, please.

**Comment by ike709** [comment]
Conflicts fixed.

Also isn't the point of TotalPositions to keep track of the total available positions...?

**Comment by Zumorica** [comment]
> Also isn't the point of TotalPositions to keep track of the total available positions...?

Kinda. Ideally, prototypes shouldn't be modified like that. We should implement a system that keeps track of job positions and loads the initial total available positions from TotalPositions in job prototypes.

**Comment by RemieRichards** [comment]
No, it's to define the maximum available positions.
If you look at AssignJobs() it doesn't reduce the positions in the prototype, it keeps track of them separately.

**Comment by RemieRichards** [comment]
It looks like `GetAvailablePositions()` is what you want, it gets a dict of available jobs, discounting those that are out of slots (all spawned in). Right now you're just enumerating all jobs, you should use `GetAvailablePositions()` instead and then do `AddSpawnedPosition(jobid)` when you latejoin someone in, to keep the book-keeping working.

This should definitely be done before merger, as there's a working system there already.

File: Content.Server/GameTicking/GameTickerCommands.cs [hunk 1]
```diff
@@ -192,8 +204,22 @@ public void Execute(IConsoleShell shell, IPlayerSession player, string[] args)
                 shell.SendText(player, "Round has not started.");
                 return;
             }
+            else if(ticker.RunLevel == GameRunLevel.InRound)
+            {
+                string ID = args[0];
+                var positions = ticker.GetAvailablePositions();
+
+                if(positions.GetValueOrDefault(ID, 0) == 0)
```
Comment (RemieRichards) [inline Content.Server/GameTicking/GameTickerCommands.cs:212]: I'd leave a comment that the `== 0` is intentional, ie: `// n < 0 is treated as infinite`

Comment (ike709) [reply Content.Server/GameTicking/GameTickerCommands.cs]: Addressed.

**Comment by ike709** [comment]
@Zumorica @RemieRichards Addressed. Also fixed the icons that got changed in a recently merged PR.

`AddSpawnedPosition(jobid)` isn't necessary because it's already called in `SpawnPlayer()`

**Comment by ike709** [comment]
Tested it and it all still works, including not letting you latejoin as a job with no available slots.

File: Content.Client/UserInterface/LateJoinGui.cs [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+                            Children =
+                            {
+                                jobList
+                            }
+                        }
+                    }
+                };
+            Contents.AddChild(vBox);
+
+            foreach (var job in _prototypeManager.EnumeratePrototypes<JobPrototype>().OrderBy(j => j.Name))
```
Comment (RemieRichards) [inline Content.Client/UserInterface/LateJoinGui.cs:52]: Is it intentional to show all jobs here? even if they're not available?

Might want to do 1 AvailablePositions() call before this loop, and then filter out jobs here using .ContainsKey() on the available dict.

Comment (ike709) [reply Content.Client/UserInterface/LateJoinGui.cs:52]: See discord.

Comment (RemieRichards) [reply Content.Client/UserInterface/LateJoinGui.cs:52]: Ok. be about half an hour/forty five mins cos traffic.

**Pull request merged**